### PR TITLE
Multi-Region CMS Support (hot standby / pilot light)

### PR DESCRIPTION
**Pull request by fieldju**
- Modified various commands to support being used outside the primary region.
- Added new composite command that can stand up the necessary resources in a secondary region of an existing environment to have a hot standby / pilot light CMS stack.

[*EDIT*]: for people that end up here, the new composite command updates an existing environment to be like so.

image

### CODE REVIEW

File: src/main/java/com/nike/cerberus/operation/core/WhitelistCidrForVpcAccessOperation.java [hunk 1]
```diff
@@ -105,11 +108,11 @@ public void run(WhitelistCidrForVpcAccessCommand command) {
 
     @Override
     public boolean isRunnable(WhitelistCidrForVpcAccessCommand command) {
+        Regions region = command.getStackRegion().orElse(configStore.getPrimaryRegion());
         try {
-            cloudFormationService.getStackId(configStore.getPrimaryRegion(), Stack.IAM_ROLES.getFullName(environmentName));
+            cloudFormationService.getStackId(region, Stack.SECURITY_GROUPS.getFullName(environmentName));
         } catch (IllegalArgumentException iae) {
```
Comment (melanahammel) [inline src/main/java/com/nike/cerberus/operation/core/WhitelistCidrForVpcAccessOperation.java]: In another exception above (https://github.com/Nike-Inc/cerberus-lifecycle-cli/pull/134/files#diff-35cf96ae7b57f8594ef82007bac0c7e9R126) IAE was changed to a general Exception for the same call, getStackID(). Should the same happen here?

File: src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java [hunk 2]
```diff
@@ -47,18 +39,20 @@
...
-import java.util.List;
+import java.util.*;
 
-public class EnvironmentConfigToArgsMapper {
+import static com.nike.cerberus.domain.cloudformation.CloudFormationParametersDelegate.STACK_REGION;
 
-    public static final String STACK_NAME_KEY = "--stack-name";
+public class EnvironmentConfigToArgsMapper {
 
     private EnvironmentConfigToArgsMapper() {
```
[the remark this answers is not in the document]
Comment (fieldju) [reply src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: That's not code I edited, but it's private because this is a utils class meant to be used via it's static methods and not instantiated.

[the remark this answers is not in the document]
Comment (fieldju) [reply src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: That's not code I edited, but it's private because this is a utils class meant to be used via it's static methods and not instantiated.

File: src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java [hunk 3]
```diff
@@ -86,13 +80,36 @@ private EnvironmentConfigToArgsMapper() {
...
     }
 
     public static List<String> getArgsForCommand(EnvironmentConfig environmentConfig, String commandName, String[] passedArgs) {
         if (environmentConfig == null) {
             return Lists.newArrayList(passedArgs);
         }
+        return getArgsForCommand(environmentConfig, commandName, passedArgs, environmentConfig.getPrimaryRegion());
+    }
+
+    public static List<String> getArgsForCommand(EnvironmentConfig environmentConfig, String commandName, String[] passedArgs, Optional<String> region) {
```
[the remark this answers is not in the document]
Comment (fieldju) [reply src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: This one optionally takes the region and defaults to the primary region where as the other explicitly expects the region to be passed.

File: src/main/java/com/nike/cerberus/domain/cloudformation/CloudFormationParametersDelegate.java [hunk 4]
```diff
@@ -29,6 +32,9 @@
     public static final String TAG_LONG_ARG = "--TAG";
     public static final String TAG_SHORT_ARG = "-T";
 
+    public static final String STACK_REGION = "--stack-region";
+    public static final String STACK_REGION_DESCRIPTION = "The region for the stack, ex: us-west-2, will default to the primary region.";
```
[the remark this answers is not in the document]
Comment (fieldju) [reply src/main/java/com/nike/cerberus/domain/cloudformation/CloudFormationParametersDelegate.java]: It's meant to communicate that if the optionally flag is not supplied that the primary region will be used.

File: src/main/java/com/nike/cerberus/operation/core/WhitelistCidrForVpcAccessOperation.java (hunk 1 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/operation/core/WhitelistCidrForVpcAccessOperation.java]: I think so, the underlying exception type changed in the AWS ask at some point.

- [ ] update exception type.

File: src/main/java/com/nike/cerberus/operation/composite/CompositeOperation.java [hunk 5]
```diff
@@ -115,6 +115,7 @@ public void run(T compositeCommand) {
         }
     }
 
+
```
[the remark this answers is not in the document]
Comment (fieldju) [reply src/main/java/com/nike/cerberus/operation/composite/CompositeOperation.java]: - [ ] get rid of whitespace.

File: src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (melanahammel) [reply src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: Thanks for explaining that!

File: src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (melanahammel) [reply src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: But if both of those fourth arguments are Strings, which function will be called in this example?

.. = getArgsForCommand(environmentConfig, commandName, passedArgs, "us-west-2");

Or does it not matter, because the outcome will be the same either way?

File: src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java [hunk 6]
```diff
@@ -86,13 +80,36 @@ private EnvironmentConfigToArgsMapper() {
         }
 
         // convert to string array and return
-        return args.toArray(new String[args.size()]);
+        return args.toArray(new String[0]);
```
Comment (mayitbeegh) [inline src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: This is probably no necessary but doesn't hurt either.

File: src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fieldju) [reply src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: One takes a String the other takes a String Optional, those are different method signatures.

[the remark this answers is not in the document]
Comment (fieldju) [reply src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: One takes a String the other takes a String Optional, those are different method signatures.

File: src/main/java/com/nike/cerberus/domain/cloudformation/CmsParameters.java [hunk 7]
```diff
@@ -38,12 +38,12 @@
     @JsonUnwrapped
     private LaunchConfigParametersDelegate launchConfigParameters = new LaunchConfigParametersDelegate();
 
-    public String getBaseStackName() {
-        return baseStackName;
+    public String getCmsIamRoleName() {
```
Comment (mayitbeegh) [inline src/main/java/com/nike/cerberus/domain/cloudformation/CmsParameters.java]: Why is this changed?

File: src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java (hunk 6 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: Intellij did that, I can revert it.

Comment (mayitbeegh) [reply src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: I don't mind either way

File: src/main/java/com/nike/cerberus/domain/cloudformation/CmsParameters.java (hunk 7 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/domain/cloudformation/CmsParameters.java]: You can not do x-region stack reference imports, so we have to explicitly provide via a parameter the thing (the CMS IAM role) we used to import in CloudFormation from the base stack.

Comment (fieldju) [reply src/main/java/com/nike/cerberus/domain/cloudformation/CmsParameters.java]: IAM roles are global but created in the primary region and we want to use said iam role in the secondary region CMS stack.

File: src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java (hunk 6 above)
Comment (fieldju) [reply src/main/java/com/nike/cerberus/cli/EnvironmentConfigToArgsMapper.java]: - [ ] revert this.
